File: experiments/insn_decode_test.py
"""
insn_decode_test.py - Instruction decoding experiment for IDA Python

Tests:
- decode_insn() API
- insn_t structure (itype, ea, size, ops[])
- op_t operand structure (type, dtype, reg, addr, value)
- itype to mnemonic mapping

Run with: ask_ida exec file experiments/insn_decode_test.py
"""

import idaapi
import idautils
import idc
import ida_funcs
import ida_ua
import ida_bytes
import ida_name

# Try to import ida_allins for instruction type constants
try:
    import ida_allins
    HAS_ALLINS = True
except ImportError:
    HAS_ALLINS = False
    print("[!] ida_allins not available - itype constants won't be resolved")

# Operand type constants and names
OPTYPE_NAMES = {
    ida_ua.o_void: "o_void",      # No operand
    ida_ua.o_reg: "o_reg",        # Register (al, ax, es, etc)
    ida_ua.o_mem: "o_mem",        # Direct memory reference
    ida_ua.o_phrase: "o_phrase",  # Indirect [reg]
    ida_ua.o_displ: "o_displ",    # Indirect [reg+disp]
    ida_ua.o_imm: "o_imm",        # Immediate value
    ida_ua.o_far: "o_far",        # Far code reference
    ida_ua.o_near: "o_near",      # Near code reference
}

# Add processor-specific types
for i in range(6):
    key = getattr(ida_ua, f"o_idpspec{i}", None)
    if key is not None:
        OPTYPE_NAMES[key] = f"o_idpspec{i}"

# Data type names
DTYPE_NAMES = {
    ida_ua.dt_byte: "dt_byte",       # 8 bit
    ida_ua.dt_word: "dt_word",       # 16 bit
    ida_ua.dt_dword: "dt_dword",     # 32 bit
    ida_ua.dt_float: "dt_float",     # 4 byte float
    ida_ua.dt_double: "dt_double",   # 8 byte float
    ida_ua.dt_tbyte: "dt_tbyte",     # 10 byte float
    ida_ua.dt_qword: "dt_qword",     # 64 bit
    ida_ua.dt_byte16: "dt_byte16",   # 128 bit
    ida_ua.dt_fword: "dt_fword",     # 48 bit
}


def decode_instruction(ea):
    """
    Decode an instruction at the given address.

    Returns dict with:
    - ea: Address
    - itype: Instruction type code
    - size: Instruction size in bytes
    - mnemonic: Mnemonic string
    - disasm: Full disassembly line
    - operands: List of operand dicts
    """
    insn = ida_ua.insn_t()
    length = ida_ua.decode_insn(insn, ea)

    if length <= 0:
        return None

    result = {
        'ea': insn.ea,
        'itype': insn.itype,
        'size': insn.size,
        'mnemonic': idc.print_insn_mnem(ea),
        'disasm': idc.generate_disasm_line(ea, 0),
        'operands': []
    }

    # Process operands
    for i in range(ida_ua.UA_MAXOP):
        op = insn.ops[i]
        if op.type == ida_ua.o_void:
            break

        op_info = {
            'n': i,
            'type': op.type,
            'type_name': OPTYPE_NAMES.get(op.type, f"unknown({op.type})"),
            'dtype': op.dtype,
            'dtype_name': DTYPE_NAMES.get(op.dtype, f"unknown({op.dtype})"),
            'reg': op.reg,
            'addr': op.addr,
            'value': op.value,
            'text': idc.print_operand(ea, i),
        }
        result['operands'].append(op_info)

    return result


def print_instruction(insn_info):
    """Pretty print decoded instruction info."""
    if insn_info is None:
        print("  [Failed to decode]")
        return

    print(f"  Address:  0x{insn_info['ea']:x}")
    print(f"  itype:    {insn_info['itype']}")
    print(f"  Size:     {insn_info['size']} bytes")
    print(f"  Mnemonic: {insn_info['mnemonic']}")
    print(f"  Disasm:   {insn_info['disasm']}")
    print(f"  Operands: {len(insn_info['operands'])}")

    for op in insn_info['operands']:
        print(f"    [{op['n']}] type={op['type_name']} dtype={op['dtype_name']} text='{op['text']}'")
        if op['type'] == ida_ua.o_reg:
            print(f"        reg={op['reg']}")
        elif op['type'] in (ida_ua.o_mem, ida_ua.o_near, ida_ua.o_far):
            print(f"        addr=0x{op['addr']:x}")
        elif op['type'] == ida_ua.o_imm:
            print(f"        value=0x{op['value']:x} ({op['value']})")
        elif op['type'] in (ida_ua.o_displ, ida_ua.o_phrase):
            print(f"        reg={op['reg']} addr=0x{op['addr']:x}")


def build_itype_mapping():
    """
    Build a mapping from itype values to mnemonic strings.

    This scans all instructions in the database to discover
    which itype values correspond to which mnemonics.
    """
    itype_to_mnem = {}

    func_count = ida_funcs.get_func_qty()
    print(f"Scanning {func_count} functions for itype mapping...")

    for i in range(func_count):
        func = ida_funcs.getn_func(i)
        if not func:
            continue

        for ea in idautils.FuncItems(func.start_ea):
            insn = ida_ua.insn_t()
            if ida_ua.decode_insn(insn, ea) > 0:
                if insn.itype not in itype_to_mnem:
                    mnem = idc.print_insn_mnem(ea)
                    itype_to_mnem[insn.itype] = mnem

    return itype_to_mnem


def analyze_itype_distribution():
    """Analyze instruction type distribution across the database."""
    itype_stats = {}  # itype -> (mnemonic, count)

    func_count = ida_funcs.get_func_qty()
    print(f"Analyzing instruction distribution in {func_count} functions...")

    total_insns = 0
    for i in range(func_count):
        func = ida_funcs.getn_func(i)
        if not func:
            continue

        for ea in idautils.FuncItems(func.start_ea):
            insn = ida_ua.insn_t()
            if ida_ua.decode_insn(insn, ea) > 0:
                total_insns += 1
                if insn.itype not in itype_stats:
                    mnem = idc.print_insn_mnem(ea)
                    itype_stats[insn.itype] = [mnem, 0]
                itype_stats[insn.itype][1] += 1

    # Sort by count
    sorted_stats = sorted(itype_stats.items(), key=lambda x: x[1][1], reverse=True)

    print(f"\nTotal instructions analyzed: {total_insns}")
    print(f"Unique instruction types: {len(itype_stats)}")
    print(f"\nTop 30 instructions by frequency:")
    print(f"{'itype':<8} {'mnemonic':<12} {'count':<10} {'%':>6}")
    print("-" * 40)

    for itype, (mnem, count) in sorted_stats[:30]:
        pct = (count / total_insns) * 100 if total_insns > 0 else 0
        print(f"{itype:<8} {mnem:<12} {count:<10} {pct:>5.1f}%")

    return itype_stats


def find_operand_type_examples():
    """Find example instructions for each operand type."""
    examples = {}  # optype -> instruction info

    func_count = ida_funcs.get_func_qty()

    for i in range(func_count):
        if len(examples) >= len(OPTYPE_NAMES):
            break

        func = ida_funcs.getn_func(i)
        if not func:
            continue

        for ea in idautils.FuncItems(func.start_ea):
            insn_info = decode_instruction(ea)
            if insn_info:
                for op in insn_info['operands']:
                    optype = op['type']
                    if optype not in examples and optype != ida_ua.o_void:
                        examples[optype] = insn_info

    return examples


def generate_itype_header():
    """Generate a C++ header with itype to mnemonic mapping."""
    mapping = build_itype_mapping()

    print(f"\n// Generated itype mapping ({len(mapping)} entries)")
    print("static const char* itype_to_mnemonic[] = {")

    max_itype = max(mapping.keys()) if mapping else 0
    for i in range(max_itype + 1):
        if i in mapping:
            print(f'    /* {i:4} */ "{mapping[i]}",')
        else:
            print(f'    /* {i:4} */ nullptr,')

    print("};")
    print(f"static const size_t itype_count = {max_itype + 1};")


def run_tests():
    """Run all instruction decoding tests."""
    print("=" * 60)
    print("INSTRUCTION DECODING TEST")
    print("=" * 60)

    # Test 1: Decode first instruction of some functions
    print("\n--- Test 1: First instruction of functions ---\n")
    func_count = ida_funcs.get_func_qty()
    shown = 0

    for i in range(func_count):
        if shown >= 5:
            break
        func = ida_funcs.getn_func(i)
        if not func:
            continue

        func_name = ida_name.get_name(func.start_ea)
        print(f"Function: {func_name}")

        insn_info = decode_instruction(func.start_ea)
        print_instruction(insn_info)
        print()
        shown += 1

    # Test 2: Operand type examples
    print("\n--- Test 2: Operand type examples ---\n")
    examples = find_operand_type_examples()

    for optype, insn_info in sorted(examples.items()):
        print(f"Example for {OPTYPE_NAMES.get(optype, 'unknown')}:")
        print_instruction(insn_info)
        print()

    # Test 3: itype distribution
    print("\n--- Test 3: itype distribution ---\n")
    analyze_itype_distribution()

    # Test 4: Generate C++ mapping (partial output)
    print("\n--- Test 4: C++ itype mapping (preview) ---\n")
    mapping = build_itype_mapping()
    print(f"Found {len(mapping)} unique itype values")
    print("First 10 entries:")
    for itype in sorted(mapping.keys())[:10]:
        print(f"  itype {itype:4} -> {mapping[itype]}")


if __name__ == "__main__":
    run_tests()

File: experiments/insn_decode_test.cpp
/**
 * insn_decode_test.cpp - Standalone instruction decoding experiment
 *
 * Tests IDA SDK instruction decoding APIs:
 * - decode_insn() for decoding instructions
 * - insn_t structure fields (itype, ea, size, ops)
 * - op_t operand structure
 * - Mnemonic retrieval via print_insn_mnem()
 *
 * Build: Use idalib (headless IDA)
 */

#include <ida.hpp>
#include <idp.hpp>
#include <ua.hpp>
#include <bytes.hpp>
#include <funcs.hpp>
#include <name.hpp>
#include <kernwin.hpp>
#include <diskio.hpp>

#include <cstdio>
#include <string>
#include <vector>
#include <map>

// Operand type names for display
static const char* operand_type_names[] = {
    "o_void",     // 0 - No operand
    "o_reg",      // 1 - General Register
    "o_mem",      // 2 - Direct memory reference
    "o_phrase",   // 3 - Indirect [reg]
    "o_displ",    // 4 - Indirect [reg+disp]
    "o_imm",      // 5 - Immediate value
    "o_far",      // 6 - Far code reference
    "o_near",     // 7 - Near code reference
    "o_idpspec0", // 8 - Processor specific
    "o_idpspec1", // 9
    "o_idpspec2", // 10
    "o_idpspec3", // 11
    "o_idpspec4", // 12
    "o_idpspec5", // 13
};

// Get operand type name
const char* get_optype_name(optype_t type) {
    if (type < sizeof(operand_type_names)/sizeof(operand_type_names[0])) {
        return operand_type_names[type];
    }
    return "unknown";
}

// Data type (dtype) names
static const char* dtype_names[] = {
    "dt_byte",    // 0 - 8 bit
    "dt_word",    // 1 - 16 bit
    "dt_dword",   // 2 - 32 bit
    "dt_float",   // 3 - 4 byte float
    "dt_double",  // 4 - 8 byte float
    "dt_tbyte",   // 5 - 10 byte float
    "dt_packreal",// 6 - packed real
    "dt_qword",   // 7 - 64 bit
    "dt_byte16",  // 8 - 128 bit
    "dt_code",    // 9 - code pointer
    "dt_void",    // 10 - void
    "dt_fword",   // 11 - 48 bit
    "dt_bitfield",// 12 - bit field
    "dt_string",  // 13 - string
    "dt_unicode", // 14 - unicode string
    "dt_ldbl",    // 15 - long double
    "dt_byte32",  // 16 - 256 bit
    "dt_byte64",  // 17 - 512 bit
};

const char* get_dtype_name(char dtype) {
    unsigned char dt = (unsigned char)dtype;
    if (dt < sizeof(dtype_names)/sizeof(dtype_names[0])) {
        return dtype_names[dt];
    }
    return "unknown";
}

// Structure to hold decoded instruction info
struct DecodedInstruction {
    ea_t ea;                      // Linear address
    uint16 itype;                 // Instruction type code
    uint16 size;                  // Instruction size in bytes
    std::string mnemonic;         // Mnemonic string
    std::string disasm;           // Full disassembly line

    // Operand info
    struct Operand {
        optype_t type;            // Operand type (o_reg, o_mem, etc.)
        char dtype;               // Data type (dt_byte, dt_dword, etc.)
        uint16 reg;               // Register number (if o_reg)
        ea_t addr;                // Address (if o_mem, o_near, o_far)
        uint64 value;             // Immediate value (if o_imm)
        std::string text;         // Operand text representation
    };
    std::vector<Operand> operands;
};

// Decode an instruction at address
bool decode_instruction_at(ea_t ea, DecodedInstruction& out) {
    insn_t insn;
    int len = decode_insn(&insn, ea);
    if (len <= 0) {
        return false;
    }

    out.ea = insn.ea;
    out.itype = insn.itype;
    out.size = insn.size;

    // Get mnemonic
    qstring mnem;
    print_insn_mnem(&mnem, ea);
    out.mnemonic = mnem.c_str();

    // Get full disassembly
    qstring disasm;
    generate_disasm_line(&disasm, ea, 0);
    tag_remove(&disasm);
    out.disasm = disasm.c_str();

    // Process operands
    for (int i = 0; i < UA_MAXOP; i++) {
        const op_t& op = insn.ops[i];
        if (op.type == o_void) break;

        DecodedInstruction::Operand operand;
        operand.type = op.type;
        operand.dtype = op.dtype;
        operand.reg = op.reg;
        operand.addr = op.addr;
        operand.value = op.value;

        // Get operand text
        qstring op_text;
        print_operand(&op_text, ea, i);
        tag_remove(&op_text);
        operand.text = op_text.c_str();

        out.operands.push_back(operand);
    }

    return true;
}

// Print decoded instruction info
void print_instruction(const DecodedInstruction& insn) {
    printf("Address:    0x%llx\n", (unsigned long long)insn.ea);
    printf("itype:      %u\n", insn.itype);
    printf("Size:       %u bytes\n", insn.size);
    printf("Mnemonic:   %s\n", insn.mnemonic.c_str());
    printf("Disasm:     %s\n", insn.disasm.c_str());
    printf("Operands:   %zu\n", insn.operands.size());

    for (size_t i = 0; i < insn.operands.size(); i++) {
        const auto& op = insn.operands[i];
        printf("  [%zu] type=%s dtype=%s text='%s'\n",
               i,
               get_optype_name(op.type),
               get_dtype_name(op.dtype),
               op.text.c_str());

        switch (op.type) {
            case o_reg:
                printf("      reg=%u\n", op.reg);
                break;
            case o_mem:
            case o_near:
            case o_far:
                printf("      addr=0x%llx\n", (unsigned long long)op.addr);
                break;
            case o_imm:
                printf("      value=0x%llx (%llu)\n",
                       (unsigned long long)op.value,
                       (unsigned long long)op.value);
                break;
            case o_displ:
            case o_phrase:
                printf("      reg=%u addr=0x%llx\n",
                       op.reg, (unsigned long long)op.addr);
                break;
            default:
                break;
        }
    }
    printf("\n");
}

// Build itype frequency map across all functions
void analyze_itype_distribution() {
    std::map<uint16, std::pair<std::string, int>> itype_stats;  // itype -> (mnemonic, count)

    int func_count = get_func_qty();
    printf("Analyzing %d functions...\n", func_count);

    for (int i = 0; i < func_count; i++) {
        func_t* func = getn_func(i);
        if (!func) continue;

        ea_t ea = func->start_ea;
        while (ea < func->end_ea && ea != BADADDR) {
            insn_t insn;
            int len = decode_insn(&insn, ea);
            if (len > 0) {
                auto& entry = itype_stats[insn.itype];
                if (entry.first.empty()) {
                    qstring mnem;
                    print_insn_mnem(&mnem, ea);
                    entry.first = mnem.c_str();
                }
                entry.second++;
                ea += len;
            } else {
                ea = next_head(ea, func->end_ea);
            }
        }
    }

    // Sort by count and print top instructions
    std::vector<std::tuple<int, uint16, std::string>> sorted;
    for (const auto& p : itype_stats) {
        sorted.emplace_back(p.second.second, p.first, p.second.first);
    }
    std::sort(sorted.rbegin(), sorted.rend());

    printf("\nTop 30 instruction types by frequency:\n");
    printf("%-8s %-12s %s\n", "itype", "mnemonic", "count");
    printf("-------------------------------------\n");
    for (size_t i = 0; i < 30 && i < sorted.size(); i++) {
        printf("%-8u %-12s %d\n",
               std::get<1>(sorted[i]),
               std::get<2>(sorted[i]).c_str(),
               std::get<0>(sorted[i]));
    }

    printf("\nTotal unique instruction types: %zu\n", itype_stats.size());
}

// Main test function - call from IDA or idalib
void run_insn_decode_test() {
    printf("=== Instruction Decode Test ===\n\n");

    // Test 1: Decode first few instructions of each function
    printf("--- Test 1: First instruction of functions ---\n\n");

    int func_count = get_func_qty();
    int shown = 0;

    for (int i = 0; i < func_count && shown < 5; i++) {
        func_t* func = getn_func(i);
        if (!func) continue;

        qstring func_name;
        get_func_name(&func_name, func->start_ea);
        printf("Function: %s\n", func_name.c_str());

        DecodedInstruction decoded;
        if (decode_instruction_at(func->start_ea, decoded)) {
            print_instruction(decoded);
            shown++;
        }
    }

    // Test 2: Show different operand types
    printf("--- Test 2: Finding different operand types ---\n\n");

    std::map<optype_t, DecodedInstruction> optype_examples;

    for (int i = 0; i < func_count && optype_examples.size() < 8; i++) {
        func_t* func = getn_func(i);
        if (!func) continue;

        ea_t ea = func->start_ea;
        while (ea < func->end_ea && ea != BADADDR) {
            DecodedInstruction decoded;
            if (decode_instruction_at(ea, decoded)) {
                for (const auto& op : decoded.operands) {
                    if (optype_examples.find(op.type) == optype_examples.end()) {
                        optype_examples[op.type] = decoded;
                    }
                }
            }
            ea = next_head(ea, func->end_ea);
        }
    }

    printf("Found examples for %zu operand types:\n\n", optype_examples.size());
    for (const auto& p : optype_examples) {
        printf("Operand type %s:\n", get_optype_name(p.first));
        print_instruction(p.second);
    }

    // Test 3: itype distribution
    printf("--- Test 3: Instruction type distribution ---\n\n");
    analyze_itype_distribution();
}

// Entry point for standalone testing
#ifdef STANDALONE_TEST
int main(int argc, char* argv[]) {
    if (argc < 2) {
        printf("Usage: %s <idb_path>\n", argv[0]);
        return 1;
    }

    // Initialize idalib
    if (!init_library()) {
        printf("Failed to initialize idalib\n");
        return 1;
    }

    // Open database
    if (!open_database(argv[1], false)) {
        printf("Failed to open database: %s\n", argv[1]);
        term_library();
        return 1;
    }

    // Run tests
    run_insn_decode_test();

    // Cleanup
    close_database(false);
    term_library();

    return 0;
}
#endif
